### ISSUE TITLE
podman makes two `getpwnam_r(3)` calls during startup to look for `buildah` and `storage` users

### ISSUE THREAD

**Issue by kelvie-pstg** [containers/storage#1227]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

There are two places where podman calls https://pkg.go.dev/os/user#Lookup during static initialization:

https://github.com/containers/podman/blob/bdaac4b2b6573d0954c58da9968536017160b1ed/vendor/github.com/containers/storage/pkg/chrootarchive/archive.go#L23
https://github.com/containers/podman/blob/bdaac4b2b6573d0954c58da9968536017160b1ed/vendor/github.com/containers/buildah/copier/copier.go#L48

 On linux (with CGO enabled, which is the default), this results in a call to getpwnam_r, which on systems with LDAP installed and enabled via `/etc/nsswitch.conf`:

```
passwd:    files ldap
```

This results in two calls to the LDAP server to look for the users `buildah` and `storage`.

**Steps to reproduce the issue:**

1. Set up /etc/ldap.conf and enable it in /etc/nsswitch.conf using `passwd: files ldap`

2. Call podman --help

3. Notice that network calls are made to the LDAP server to look for the `buildah` and `storage` users

**Describe the results you received:**

Slow LDAP servers cause this call to be really slow, and if the network is down, podman becomes *really* slow.

**Describe the results you expected:**

No extraneous calls to LDAP servers or getpwnam_r unless necessary.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
     root@vm-kwong2:~# podman version
Version:      3.2.1
API Version:  3.2.1
Go Version:   go1.13.8
Built:        Wed Dec 31 16:00:00 1969
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.21.3
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - pids
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/bin/conmon'
    path: /usr/bin/conmon
    version: 'conmon version 2.0.25, commit: unknown'
  cpus: 8
  distribution:
    distribution: purity
    version: "20.04"
  eventLogger: journald
  hostname: vm-kwong2
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.4.114+
  linkmode: dynamic
  memFree: 666615808
  memTotal: 25088991232
  ociRuntime:
    name: runc
    package: 'runc: /usr/sbin/runc'
    path: /usr/sbin/runc
    version: |-
      runc version 1.0.0~rc95-0ubuntu1~20.04.2
      spec: 1.0.2-dev
      go: go1.13.8
      libseccomp: 2.5.1
  os: linux
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 0
  swapTotal: 0
  uptime: 189h 6m 57.7s (Approximately 7.88 days)
registries: {}
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 5
    paused: 0
    running: 5
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 5
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.2.1
  Built: 0
  BuiltTime: Wed Dec 31 16:00:00 1969
  GitCommit: ""
  GoVersion: go1.13.8
  OsArch: linux/amd64
  Version: 3.2.1

```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman/now 3.2.1+ds1-2ubuntu2+pure1 amd64 [installed,local]
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

No

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical device

**Comment by giuseppe** [comment]
I had a look at the issue but I don't see any valid workaround for that.

Would it work if we move this code before the re-exec so its cost is limited to the cases where we really need it?

**Comment by rhatdan** [comment]
```
	// Attempt a user and host lookup to force libc (glibc, and possibly others that use dynamic
	// modules to handle looking up user and host information) to load modules that match the libc
	// our binary is currently using.  Hopefully they're loaded on first use, so that they won't
	// need to be loaded after we've chrooted into the rootfs, which could include modules that
	// don't match our libc and which can't be loaded, or modules which we don't want to execute
	// because we don't trust their code.
	_, _ = user.Lookup("buildah")
```
Would this be better if we just looked up the root user?  Rather then a non existing user?  I have no idea the history of this code BTW, (Don't believe I added it.)

**Comment by rhatdan** [comment]
@nalind PTAL

**Comment by giuseppe** [comment]
opened two PRs:

https://github.com/containers/storage/pull/1228
https://github.com/containers/buildah/pull/3966

**Comment by kelvie-pstg** [comment]
> ```
> 	// Attempt a user and host lookup to force libc (glibc, and possibly others that use dynamic
> 	// modules to handle looking up user and host information) to load modules that match the libc
> 	// our binary is currently using.  Hopefully they're loaded on first use, so that they won't
> 	// need to be loaded after we've chrooted into the rootfs, which could include modules that
> 	// don't match our libc and which can't be loaded, or modules which we don't want to execute
> 	// because we don't trust their code.
> 	_, _ = user.Lookup("buildah")
> ```
> 
> Would this be better if we just looked up the root user? Rather then a non existing user? I have no idea the history of this code BTW, (Don't believe I added it.)

Yes, it would be a lot better if both of these places looked up the `root` user instead, as it wouldn't cause an LDAP lookup in our case for two seemingly random users, `storage`, and `buildah` (possibly triggering an IPS alert or something).

**Comment by kelvie-pstg** [comment]
@rhatdan  Is there another issue tracking the `buildah` lookup that's still in static init?

https://github.com/containers/podman/blob/bdaac4b2b6573d0954c58da9968536017160b1ed/vendor/github.com/containers/buildah/copier/copier.go#L48

And can I request that they both be changed to a user that is more likely to exist to avoid an ldap lookup (like `root`)

**Comment by kelvie-pstg** [comment]
I still think it's strange to have an NSS/LDAP lookup for seemingly random users coming from podman, it's sure to raise some eyebrows in our customers' environments if they catch it on the network.

**Comment by nalind** [comment]
I agree, it is.  The specific before-chrooting calls we make attempt to ensure two things: that the nsswitch configuration is initialized based on the outside-of-the-chroot /etc/nsswitch.conf file, and that the various nsswitch modules listed in that configuration are opened before we chroot into a different filesystem tree.  A module is loaded from a predictable path, and if we trigger an attempt to load one after we've chrooted, bad things can happen.

The first part (loading nsswitch.conf configuration) happens as soon as we attempt a user or host lookup; I think either one, with any input, would be sufficient on its own.

That second part (loading modules), though, happens only for a given module when it needs to be called into, and passwd lookups tend to be configured to only consult a module after any/all modules that the configuration says should be consulted before it have failed to answer a query.  If we looked up a passwd entry for a user that we were confident was known to nss_files, nss_ldap (or any other modules that are configured to be consulted after nss_files) wouldn't necessarily be loaded before the chroot(), so if we attempted a lookup after chrooting that the loaded nss_files module couldn't satisfy, we'd end up trying to open other modules which hadn't been consulted yet, which would be re-creating the problem that all of this is trying to avoid.

The short version is that looking up "root" would create problems, and we kind of depend on "storage" and "buildah" not being valid user names, which seems kind of delicate to me.

(For reference, I tried changing the user names we look up and used `podman run -d fedora sh -c 'touch /13; chown 13:13 /13'` followed by `strace -f -e chroot,open,openat podman cp ${containerID}:13 /tmp` to spot-check the effects.  UID 13 and GID 13 are not present in my /etc/passwd and /etc/group files.)
